### PR TITLE
[FIX] http: logout when a database is deleted

### PR DESCRIPTION
**Pull request by Xavier-Do**
When an OperationalError error occurs trying to initialize a registry,
a possible reason is that the database does not exist, this is the main
reason why 28581cc73442747463386b1a84d1c54d15128078 introduced a logout.

Since #49111, if the error occurs on a /web page, the error is raised,
get_response is never reached, meaning that the session is not saved.

This will lead to an error everytime the page is refreshed on /web since
the database won't be removed from the session.

- connect to a database
- drop it from the database manager
- access /web. An ISE occurs

This is partially fixed in #82874, using logout on the session dropping
the database but the problem will remain for other sessions.

This commit proposes to save the session in all case when logout is
called. Looks like it makes sense in this case since performance-wise
logout won't be called much, and we don't want to lose the logout
information if we don't reach get_response().

Another solution would be to save the session only before raising.

This is a partial alternative to #54102

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Xavier-Do** [comment]
Waiting for #82874 to be merged to add a test for this use case.

**Review by Julien00859 — APPROVED** [review]

**Comment by xmo-odoo** [comment]
@Xavier-Do  #82874  is merged, however should this fix not be merged in 13.0? Or do you intend to backport it later?

Also is there still a point to #54102 once this is merged?

**Comment by Xavier-Do** [comment]
> Xavier-Do #82874 is merged, however should this fix not be merged in 13.0? Or do you intend to backport it later?
I don't know. We can talk about that. I will write some tests in master at least.

>Also is there still a point to #54102 once this is merged?
I'm not 100% sure but i don't think so. Maybe if the code base changes and the db is not migrated?

File: odoo/tests/common.py
Comment (xmo-odoo) [inline odoo/tests/common.py:724]: Since this is moved, it should be folded into a single `re.search`, compiling to a local variable is useless (either that or the `csrf_token_re` should be moved to a global)

File: addons/web/tests/test_db_manager.py
Comment (xmo-odoo) [inline addons/web/tests/test_db_manager.py:118]: What's the purpose of setting on `self`?

File: addons/web/tests/test_db_manager.py
Comment (xmo-odoo) [inline addons/web/tests/test_db_manager.py:153]: Maybe `_wait_remaining_requests` could be converted to a free function, to which a logger would be passed, for less ugly? And that way `self._logger` wouldn't be *necessary* (though it could still be useful).

File: addons/web/tests/test_db_manager.py
Comment (xmo-odoo) [inline addons/web/tests/test_db_manager.py:145]: Should be part of the commit message rather than a comment.

**Review by xmo-odoo — CHANGES_REQUESTED** [review]

File: addons/web/tests/test_db_manager.py
Comment (Xavier-Do) [reply addons/web/tests/test_db_manager.py:153]: Yes I thinked about that but I didn't want to change the behaviour or using a specific logger giving a clear information on the test causing the error.

And having a logger like that on the Base Case looked like a good idea. I agree with the `( it could still be useful).`

I still extracted _wait_remaining_requests outside HTTPCase with an optionnal logger parametter, not sure it is better like that. My use case to use it outside an HTTPCase is not usual.

File: addons/web/tests/test_db_manager.py
Comment (Xavier-Do) [reply addons/web/tests/test_db_manager.py:145]: Or both a commit message and a comment ;)

File: addons/web/tests/test_db_manager.py
Comment (xmo-odoo) [reply addons/web/tests/test_db_manager.py]: > I still extracted `_wait_remaining_requests` outside HTTPCase with an optionnal logger parametter, not sure it is better like that. My use case to use it outside an HTTPCase is not usual.

I think it's better, strictly speaking `_wait_remaining_requests` doesn't depend on any part of the test case and while it's definitely test-related (mostly) it makes sense as a separate utility. Maybe remove the `_` prefix as well?

**Comment by Xavier-Do** [comment]
Will be resurrected if needed look like it was fixed

**Pull request closed without merging**
